hermes/bin/src/runtime_extensions/hermes/sqlite/state.rs

```diff
@@ -0,0 +1,108 @@
+// cspell: words mapref
+
+//! Internal state implementation for the `SQLite` module.
+
+use std::collections::HashMap;
+
+use dashmap::{mapref::one::RefMut, DashMap};
+use once_cell::sync::Lazy;
+
+use crate::app::HermesAppName;
+
+/// Represents an individual state for a particular object.
+#[derive(Debug)]
+pub(crate) struct ResourceObjectState {
+    /// A map holding key-value pairs of an object ID and a value.
+    id_map: HashMap<u32, usize>,
+    /// The current incremental state of ID.
+    current_id: Option<u32>,
+}
+
+/// Represents the state of resources.
+pub(crate) struct ResourceState {
+    /// The state of database object.
+    db_state: ResourceObjectState,
+    /// The state of database statement object.
+    stmt_state: ResourceObjectState,
+}
+
+impl ResourceObjectState {
+    /// Create a new `ResourceObjectState` with initial state.
+    fn new() -> Self {
+        Self {
+            id_map: HashMap::new(),
+            current_id: None,
+        }
+    }
+
+    /// Adds a value into the resource. If it does not exist, allocate one and returns the
+    /// new created key ID. In case of the key ID is running out of numbers, returns
+    /// `None`.
+    pub(super) fn allocate_object(&mut self, object_ptr: usize) -> Option<u32> {
```

Comment:
Not a correct name for the function, because you are not actually allocating something init.
So it could mislead in the future.

Reply:
I'll change this to `add_object`, should be a more straightforward name for this function. As I also think that `allocate_object` is more related about low-level memory stuffs.

Reply:
I would rather we don't start doing this work for now.  its not necessary and just creates more work when we will probably need to refactor state management at a later date.  The simple implementation ehre is fine for now.